Log posting with max text length fails

##### Detailed steps causing the problem:
- Compose a log for a gc.com cache
- Enter as much text as possible as log text (until no further characters are accepted)
- Try to send the log online

##### Actual behavior after performing these steps:
Logging immediately fails with the corresponding c:geo failure window.
No long posted online.

##### Expected behavior after performing these steps:
Max characters allowed for log in c:geo seems not to match what gc.com supports (I assume).
As soon as I reduce the log text by four characters the logging works smooth.


##### Version of c:geo used:
2019.03.17-RC (current `release` branch)

##### Is the problem reproducible for you?
Yes

##### System information:
Android 5.0.1, Samsung Galaxy S4

##### Other comments and remarks:
Seems **not** related to recent logging change #7388

Still reproducible with 2020.09.22-NB, Samsung S20, Android 10:

```
09-23 15:08:18.128 31593 13621 E cgeo    : [AsyncTask #7] Error posting log
09-23 15:08:18.128 31593 13621 E cgeo    : java.lang.RuntimeException: java.io.IOException: request was not successful
09-23 15:08:18.128 31593 13621 E cgeo    : 	at io.reactivex.rxjava3.internal.util.ExceptionHelper.wrapOrThrow(ExceptionHelper.java:46)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at io.reactivex.rxjava3.internal.observers.BlockingMultiObserver.blockingGet(BlockingMultiObserver.java:94)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at io.reactivex.rxjava3.core.Single.blockingGet(Single.java:3645)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at cgeo.geocaching.connector.gc.GCWebAPI.postLog(GCWebAPI.java:459)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at cgeo.geocaching.connector.gc.GCLoggingManager.postLog(GCLoggingManager.java:137)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at cgeo.geocaching.log.LogCacheActivity$Poster.doInBackgroundInternal(LogCacheActivity.java:881)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at cgeo.geocaching.log.LogCacheActivity$Poster.doInBackgroundInternal(LogCacheActivity.java:865)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at cgeo.geocaching.utils.AbstractAsyncTaskWithProgress.doInBackground(AbstractAsyncTaskWithProgress.java:134)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at android.os.AsyncTask$3.call(AsyncTask.java:378)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at java.lang.Thread.run(Thread.java:919)
09-23 15:08:18.128 31593 13621 E cgeo    : Caused by: java.io.IOException: request was not successful
09-23 15:08:18.128 31593 13621 E cgeo    : 	at cgeo.geocaching.network.Network.lambda$static$6(Network.java:621)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at cgeo.geocaching.network.-$$Lambda$Network$S1n6XPPgWcknnAjXJlYG-giT7W4.apply(Unknown Source:2)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at io.reactivex.rxjava3.internal.operators.single.SingleFlatMap$SingleFlatMapCallback.onSuccess(SingleFlatMap.java:77)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at io.reactivex.rxjava3.internal.operators.single.SingleCreate$Emitter.onSuccess(SingleCreate.java:68)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at cgeo.geocaching.utils.RxOkHttpUtils$1.onResponse(RxOkHttpUtils.java:48)
09-23 15:08:18.128 31593 13621 E cgeo    : 	at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519)
09-23 15:08:18.128 31593 13621 E cgeo    : 	... 3 more
```

--- a/main/res/layout/logcache_activity.xml
+++ b/main/res/layout/logcache_activity.xml
@@ -51,6 +51,16 @@
                 android:minLines="5"
                 tools:text="The log text is limited to 4000 characters."/>
 
+            <TextView
+                android:id="@+id/log_characters_counter"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_marginRight="10dip"
+                android:layout_gravity="right"
+                android:visibility="gone"
+                android:textColor="?text_color"
+                android:textSize="14sp" />
+
             <LinearLayout
                 android:id="@+id/log_password_box"
                 android:layout_width="fill_parent"

--- a/main/res/values/strings.xml
+++ b/main/res/values/strings.xml
@@ -1025,6 +1025,7 @@
     <string name="cache_menu_streetview">Street View</string>
     <string name="cache_menu_browser">Open in browser</string>
     <string name="cache_menu_visit">Log Visit</string>
+    <string name="cache_log_too_long">c:geo couldn\'t submit your log cause it is too long. But thanks for trying to write such an extensive log ;-)</string>
     <string name="cache_menu_visit_offline">Quick offline log</string>
     <string name="cache_menu_spoilers">Spoiler images</string>
     <string name="cache_menu_around">Caches around</string>

--- a/main/src/cgeo/geocaching/log/LogCacheActivity.java
+++ b/main/src/cgeo/geocaching/log/LogCacheActivity.java
@@ -36,14 +36,19 @@ import cgeo.geocaching.utils.CalendarUtils;
 import cgeo.geocaching.utils.CollectionStream;
 import cgeo.geocaching.utils.ContextLogger;
 import cgeo.geocaching.utils.Log;
+import cgeo.geocaching.utils.TextUtils;
 import cgeo.geocaching.utils.ViewUtils;
 
 import android.R.string;
+import android.annotation.SuppressLint;
 import android.app.Activity;
 import android.content.Context;
 import android.content.Intent;
+import android.graphics.Color;
 import android.os.AsyncTask;
 import android.os.Bundle;
+import android.text.Editable;
+import android.text.TextWatcher;
 import android.view.Menu;
 import android.view.MenuItem;
 import android.view.View;
@@ -55,6 +60,7 @@ import android.widget.ImageView;
 import android.widget.LinearLayout;
 import android.widget.ListView;
 import android.widget.TextView;
+import android.widget.Toast;
 
 import androidx.annotation.NonNull;
 
@@ -74,6 +80,7 @@ import org.apache.commons.lang3.StringUtils;
 public class LogCacheActivity extends AbstractLoggingActivity {
 
     private static final String SAVED_STATE_LOGENTRY = "cgeo.geocaching.saved_state_logentry";
+    private static final int LOG_MAX_LENGTH = 4000;
 
     private enum SaveMode { NORMAL, FORCE, SKIP }
 
@@ -88,6 +95,8 @@ public class LogCacheActivity extends AbstractLoggingActivity {
     protected CheckBox favCheck;
     @BindView(R.id.log)
     protected EditText logEditText;
+    @BindView(R.id.log_characters_counter)
+    protected TextView logCharactersCounter;
 
     protected ImageListFragment imageListFragment;
 
@@ -223,6 +232,38 @@ public class LogCacheActivity extends AbstractLoggingActivity {
                 .setDisplayMapper(ReportProblemType::getL10n);
 
         this.imageListFragment = (ImageListFragment) getSupportFragmentManager().findFragmentById(R.id.imagelist_fragment);
+
+        // show remaining characters
+        logEditText.addTextChangedListener(new TextWatcher() {
+
+            @Override
+            public void beforeTextChanged(final CharSequence s, final int st, final int c, final int a) {
+                // do nothing
+            }
+
+            @Override
+            public void onTextChanged(final CharSequence s, final int st, final int b, final int c) {
+                // do nothing
+            }
+
+            @SuppressLint("SetTextI18n")
+            @Override
+            public void afterTextChanged(final Editable editable) {
+                final int count = TextUtils.getNormalizedStringLength(editable.toString());
+
+                if (count >= 3000) {
+                    logCharactersCounter.setVisibility(View.VISIBLE);
+                    logCharactersCounter.setText(count + "/" + LOG_MAX_LENGTH);
+                    logCharactersCounter.setTextColor(count > LOG_MAX_LENGTH ? Color.RED : getResources().getColor(Settings.isLightSkin() ? R.color.text_light : R.color.text_dark));
+
+                } else {
+                    logCharactersCounter.setVisibility(View.GONE);
+                }
+            }
+
+
+
+        });
 
         //init trackable "change all" button
         trackableActionsChangeAll.setTextView(findViewById(R.id.changebutton))
@@ -504,7 +545,11 @@ public class LogCacheActivity extends AbstractLoggingActivity {
         Log.v("LogCacheActivity.onOptionsItemSelected(" + item.getItemId() + "/" + item.getTitle() + ")");
         switch (item.getItemId()) {
             case R.id.menu_send:
-                sendLogAndConfirm();
+                if (TextUtils.getNormalizedStringLength(logEditText.getText().toString()) <= LOG_MAX_LENGTH) {
+                    sendLogAndConfirm();
+                } else {
+                    Toast.makeText(this, R.string.cache_log_too_long, Toast.LENGTH_LONG).show();
+                }
                 return true;
             case R.id.menu_image:
                 imageListFragment.startAddImageDialog();

--- a/main/src/cgeo/geocaching/utils/TextUtils.java
+++ b/main/src/cgeo/geocaching/utils/TextUtils.java
@@ -27,6 +27,7 @@ import java.util.zip.CRC32;
 
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 import org.apache.commons.lang3.StringUtils;
+import org.jetbrains.annotations.NotNull;
 
 /**
  * Misc. utils. All methods don't use Android specific stuff to use these methods in plain JUnit tests.
@@ -168,6 +169,16 @@ public final class TextUtils {
             }
         }
         return String.valueOf(chars, 0, resultSize);
+    }
+
+    /**
+     * @param str input string
+     *            As of performance reasons we non't use a REGEX here. Don't use this function for strings which could contain new-line characters like "\r\n" or "\r"
+     * @return normalized String Length like it is counted at the gc website (count UNIX new-line character "\n" as two characters)
+     */
+    public static int getNormalizedStringLength (@NotNull final String str) {
+        final String newStr = str.trim();
+        return StringUtils.countMatches(newStr, "\n") + newStr.length();
     }
 
     /**

